src/NuGet.Services.V3/CommitCollectorUtility.cs

```diff
@@ -0,0 +1,170 @@
+            }
+        }
+
+        public IEnumerable<CatalogCommitItemBatch> CreateSingleBatch(IEnumerable<CatalogCommitItem> catalogItems)
+        {
+            if (!catalogItems.Any())
+            {
+                return Enumerable.Empty<CatalogCommitItemBatch>();
+            }
+
+            var maxCommitTimestamp = catalogItems.Max(x => x.CommitTimeStamp);
+
+            return new[]
+            {
+                new CatalogCommitItemBatch(
+                    catalogItems,
+                    key: null,
+                    commitTimestamp: maxCommitTimestamp),
+            };
+        }
+
+        public List<CatalogCommitItem> GetLatestPerIdentity(IEnumerable<CatalogCommitItem> items)
+        {
+            return items
+                .GroupBy(x => x.PackageIdentity)
+                .Select(GetLatestPerSingleIdentity)
+                .ToList();
+        }
+
+        private CatalogCommitItem GetLatestPerSingleIdentity(IEnumerable<CatalogCommitItem> entries)
+        {
+            CatalogCommitItem max = null;
+            foreach (var entry in entries)
+            {
+                if (max == null)
+                {
+                    max = entry;
+                    continue;
+                }
+
+                if (!StringComparer.OrdinalIgnoreCase.Equals(max.PackageIdentity, entry.PackageIdentity))
+                {
+                    throw new InvalidOperationException("The entries compared should have the same package identity.");
+                }
+
+                if (entry.CommitTimeStamp > max.CommitTimeStamp)
+                {
+                    max = entry;
+                }
+                else if (entry.CommitTimeStamp == max.CommitTimeStamp)
+                {
+                    const string message = "There are multiple catalog leaves for a single package at one time.";
+                    _logger.LogError(
+                        message + " ID: {PackageId}, version: {PackageVersion}, commit timestamp: {CommitTimestamp:O}",
+                        entry.PackageIdentity.Id,
+                        entry.PackageIdentity.Version.ToFullString(),
+                        entry.CommitTimeStamp);
+                    throw new InvalidOperationException(message);
+                }
+            }
+
+            return max;
+        }
+
+        public ConcurrentBag<IdAndValue<IReadOnlyList<CatalogCommitItem>>> GroupById(List<CatalogCommitItem> latestItems)
+        {
+            var workEnumerable = latestItems
+                .GroupBy(x => x.PackageIdentity.Id, StringComparer.OrdinalIgnoreCase)
+                .Select(x => new IdAndValue<IReadOnlyList<CatalogCommitItem>>(x.Key, x.ToList()));
+
+            var allWork = new ConcurrentBag<IdAndValue<IReadOnlyList<CatalogCommitItem>>>(workEnumerable);
+            return allWork;
+        }
+
+        public async Task<IReadOnlyDictionary<CatalogCommitItem, PackageDetailsCatalogLeaf>> GetEntryToLeafAsync(
```

Comment:
Should we make it clear this doesn't load delete leaves? Maybe a name like `GetEntryToDetailsLeaf`? I don't feel strongly about this

Reply:
Done.